Handling FILLVAL_ATTRIBUTES Missing

sh-abinash-sh opened the pull request:
## PR Description

This PR addresses an issue in the ```_cdf.py``` file where the CDF reader would raise a KeyError if the 'FILLVAL' attribute was not present in the variable attributes of a CDF file. The fix ensures the code checks for the existence of 'FILLVAL' before accessing it, providing a more robust and error-free experience.

1. Added a conditional check to verify if 'FILLVAL' exists in attrs before attempting to use it.
If 'FILLVAL' is not present, a debug log message is generated to notify the user.
Prevents unnecessary crashes when processing CDF files that lack the 'FILLVAL' attribute.

![Screenshot 2024-12-10 100617](https://github.com/user-attachments/assets/193cbc0b-97b5-4cbb-bedc-1ba00b42be0a)


Fixes #7565

nabobalis commented:
Hello @sh-abinash-sh, thanks for the PR. Can you detail what is LLM and what you wrote when it comes to the code and the PR description? 

Also what was the code in `reproduce.py`? What did it test for?

sh-abinash-sh commented:
The reproduce.py is the code snippet which is used to reproduce the issue , i have just written it to see the issue.

sh-abinash-sh commented:
For the PR description:



> Hello @sh-abinash-sh, thanks for the PR. Can you detail what is LLM and what you wrote when it comes to the code and the PR description?
> 
> Also what was the code in `reproduce.py`? What did it test for?

I have changed the code in sunpy/sunpy/io/_cdf.py
`if np.issubdtype(data.dtype, np.floating):
      data[data == attrs['FILLVAL']] = np.nan`
    
to 


`if np.issubdtype(data.dtype, np.floating):
    if 'FILLVAL' in attrs:
        data[data == attrs['FILLVAL']] = np.nan
    else:
        log.debug(f"No 'FILLVAL' attribute found for variable {var_key}.")`

sh-abinash-sh commented:
A more detailed PR description can be :

1. Added a conditional check to verify if ```FILLVAL``` exists in ```attrs``` before attempting to use it.
2. If ```FILLVAL``` is not present, a debug log message is generated to notify the user.
3. Prevents unnecessary crashes when processing ```CDF``` files that lack the ```FILLVAL``` attribute.

nabobalis commented:
> For the PR description:
> 
> > Hello @sh-abinash-sh, thanks for the PR. Can you detail what is LLM and what you wrote when it comes to the code and the PR description?
> > Also what was the code in `reproduce.py`? What did it test for?
> 
> I have changed the code in sunpy/sunpy/io/_cdf.py `if np.issubdtype(data.dtype, np.floating): data[data == attrs['FILLVAL']] = np.nan`
> 
> to
> 
> `if np.issubdtype(data.dtype, np.floating): if 'FILLVAL' in attrs: data[data == attrs['FILLVAL']] = np.nan else: log.debug(f"No 'FILLVAL' attribute found for variable {var_key}.")`
>
> A more detailed PR description can be :
> 
> 1. Added a conditional check to verify if `FILLVAL` exists in `attrs` before attempting to use it.
> 2. If `FILLVAL` is not present, a debug log message is generated to notify the user.
> 3. Prevents unnecessary crashes when processing `CDF` files that lack the `FILLVAL` attribute.

I understand what was changed, I am asking if you made the change and wrote the PR description or was a LLM involved? 

Overall the changes are fine but we need to add a unit test to ensure that it does not break again.
There are testing guidelines here: https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html

This will also need a changelog entry, see https://github.com/sunpy/sunpy/blob/main/changelog/README.rst for guidelines. This will be a bugfix type.

sh-abinash-sh commented:
> > For the PR description:
> > > Hello @sh-abinash-sh, thanks for the PR. Can you detail what is LLM and what you wrote when it comes to the code and the PR description?
> > > Also what was the code in `reproduce.py`? What did it test for?
> > 
> > 
> > I have changed the code in sunpy/sunpy/io/_cdf.py `if np.issubdtype(data.dtype, np.floating): data[data == attrs['FILLVAL']] = np.nan`
> > to
> > `if np.issubdtype(data.dtype, np.floating): if 'FILLVAL' in attrs: data[data == attrs['FILLVAL']] = np.nan else: log.debug(f"No 'FILLVAL' attribute found for variable {var_key}.")`
> > A more detailed PR description can be :
> > 
> > 1. Added a conditional check to verify if `FILLVAL` exists in `attrs` before attempting to use it.
> > 2. If `FILLVAL` is not present, a debug log message is generated to notify the user.
> > 3. Prevents unnecessary crashes when processing `CDF` files that lack the `FILLVAL` attribute.
> 
> I understand what was changed, I am asking if you made the change and wrote the PR description or was a LLM involved?
> 
> Overall the changes are fine but we need to add a unit test to ensure that it does not break again. There are testing guidelines here: https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html
> 
> This will also need a changelog entry, see https://github.com/sunpy/sunpy/blob/main/changelog/README.rst for guidelines. This will be a bugfix type.

1. The PR description i have written with the help of Deepseek LLM, which i have corrected in later in the comment.
2. I will add the tests.
3. Also I will add changelog file.

Abinash-bit commented:
Added all the required things to merge the PR.

nabobalis commented on sunpy/io/_cdf.py:
```diff
@@ -90,7 +90,11 @@ def read_cdf(fname, **kwargs):
-                data[data == attrs['FILLVAL']] = np.nan
+                if 'FILLVAL' in attrs:
+                    data[data == attrs['FILLVAL']] = np.nan
+                else:
+                    log.debug(f"No 'FILLVAL' attribute found for variable {var_key}.")
+
```
```suggestion
```

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -23,3 +23,11 @@ def test_read_cdf():
     assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
     # Check that fillvals are replaced by NaN
     assert np.sum(np.isnan(col)) == 189
+    # Simulate missing 'FILLVAL' for one variable and ensure graceful handling
+    for ts in all_ts:
+        # Iterate over each column in the GenericTimeSeries
+        for col_name in ts.columns:
+            col_data = ts.quantity(col_name)
+            # Ensure no errors occur and data is valid
+            assert col_data is not None
+            assert np.any(np.isnan(col_data)) or np.all(~np.isnan(col_data))
```
This should be a separate unit test.

sh-abinash-sh replied:
So i should create a separate function for this?

nabobalis replied:
Yes please.

sh-abinash-sh replied:
Sure.

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -23,3 +23,15 @@ def test_read_cdf():
     assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
     # Check that fillvals are replaced by NaN
     assert np.sum(np.isnan(col)) == 189
+    # Simulate missing 'FILLVAL' for one variable and ensure graceful handling
+
+def test_generic_timeseries_columns():
+    all_ts = read_cdf(filepath)
+    for ts in all_ts:
+        # Iterate over each column in the GenericTimeSeries
+        for col_name in ts.columns:
+            col_data = ts.quantity(col_name)
+            # Ensure the data exists
+            assert col_data is not None
+            # Validate NaN handling in the column
+            assert np.any(np.isnan(col_data)) or np.all(~np.isnan(col_data))
```
```suggestion


def test_generic_timeseries_columns():
    # Simulate missing 'FILLVAL' for one variable and ensure graceful handling
    all_ts = read_cdf(filepath)
    for ts in all_ts:
        # Iterate over each column in the GenericTimeSeries
        for col_name in ts.columns:
            col_data = ts.quantity(col_name)
            # Ensure the data exists
            assert col_data is not None
            # Validate NaN handling in the column
            assert np.any(np.isnan(col_data)) or np.all(~np.isnan(col_data))
```
How does this "Simulate missing 'FILLVAL'"

sh-abinash-sh replied:
the code does not actually simulate a missing FILLVAL in the test data it verifies and handles missing data.

nabobalis replied:
So this test is actually not checking the bug is fixed?

sh-abinash-sh replied:
I have written tests so that the test function ```test_generic_timeseries_columns()``` to explicitly check for missing 'FILLVAL' and validate the behavior of the read_cdf function in such cases.

This is the code snippet:
```def test_generic_timeseries_columns():
    all_ts = read_cdf(filepath)
    for ts in all_ts:
        # Iterate over each column in the GenericTimeSeries
        for col_name in ts.columns:
            col_data = ts.quantity(col_name)
            # Ensure the data exists
            assert col_data is not None
            # Validate NaN handling in the column
            if np.issubdtype(col_data.dtype, np.floating):
                # Check for the presence of 'FILLVAL' in the column metadata
                attrs = ts.meta.get(col_name, {})
                if 'FILLVAL' in attrs:
                    # Validate that fill values are replaced with NaN
                    assert np.any(np.isnan(col_data)) or np.all(~np.isnan(col_data))
                else:
                    # If 'FILLVAL' is missing, ensure no NaN values are present
                    assert not np.any(np.isnan(col_data))
            else:
                # Non-floating-point columns should not contain NaN
                assert not np.any(np.isnan(col_data))
```


But this fails!!

![Screenshot 2024-12-11 123534](https://github.com/user-attachments/assets/7ee6b517-48bb-4d7d-81e0-5e50e274a253)

nabobalis replied:
But it fails not due to the bug we want to make sure is fixed?

Also, why is this test so complex? We just need to validate that the file has the fillval and that it's taken care of in the final timeseries. It does not need to be any fancier than that.

sh-abinash-sh replied:
Its unfortunate that , I can not understand where does the real issue lies:

Let me make a list of what i understood:

1. The current ```test_generic_timeseries_columns()``` function iterates over all columns in all GenericTimeSeries objects and checks:
       1. whether the column data exists.
       2. whether the column contains NaN values.
2. However it does not explicitly check for the presence of the FILLVAL attribute in the column metadata(attrs) which means if the FILLVAL attribute is missing than the ```read_cdf```  function should not replace any values with NaN for that column and the The test function will still pass as long as the column data is valid and does not contain unexpected NaN values. 
3. In the read_cdf function, I added a check to ensure that the 'FILLVAL' attribute exists before replacing values with NaN:
        1. If 'FILLVAL' is missing, the code skips replacing values with NaN.
        2. This means that the column data will remain unchanged, and no NaN values will be introduced.
4. How the Test Function Reacts:
         1. The column data will not contain any NaN values (since 'FILLVAL' was missing and no replacement occurred).
         2. The assertion assert ```np.any(np.isnan(col_data))``` or ```np.all(~np.isnan(col_data))``` will evaluate to True because ```np.all(~np.isnan(col_data))``` will be True (all values are not ```NaN```).
5. What Happens if 'FILLVAL' is Missing?
         1. The read_cdf function will not replace any values with NaN for that column.
         2. The column data will remain as-is, with no NaN values.
         3. The assertion assert np.any(np.isnan(col_data)) or np.all(~np.isnan(col_data)) will pass because np.all(~np.isnan(col_data)) is True.
         4. The test function will not raise any errors for this column.
6. What If the Test Function Should Explicitly Handle Missing 'FILLVAL'?

           1. Check for Floating-Point Columns
           2. Check for 'FILLVAL' Attribute
           3. Behavior for Present 'FILLVAL':
                  1. Some values are NaN (indicating that FILLVAL was replaced).
                  2. Or all values are not NaN (if no FILLVAL values were present in the data).
           4. Behavior for Missing 'FILLVAL':
                    1. No NaN values are present in the column data (since no replacement should have occurred).
           5. Non-floating-point columns should not contain NaN values, regardless of whether 'FILLVAL' is present.

sh-abinash-sh replied:
From the above understanding :

and Your question of ''How does this Simulate missing 'FILLVAL' '':

My answer is this current code :

```
def test_generic_timeseries_columns():
    all_ts = read_cdf(filepath)
    for ts in all_ts:
        for col_name in ts.columns:
            col_data = ts.quantity(col_name)
            assert col_data is not None
            assert np.any(np.isnan(col_data)) or np.all(~np.isnan(col_data))
```

Yes it handles because :
1. The column data will not contain any NaN values (since 'FILLVAL' was missing and no replacement occurred).
2. The assertion assert `np.any(np.isnan(col_data))` or `np.all(~np.isnan(col_data))` will evaluate to True because `np.all(~np.isnan(col_data))` will be True (all values are not NaN).

sh-abinash-sh replied:
Variable: EPOCH
Attributes (attrs): {'FILLVAL': np.int64(-9223372036854775808), 'LABLAXIS': 'Time', 'UNITS': 'nanoseconds', 'VALIDMIN': np.int64(-43135816000000), 'VALIDMAX': np.int64(1577880069184000000), 'SCALEMIN': np.int64(647870469184000000), 'SCALEMAX': np.int64(647956869184000000), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': 'Start of acquisition interval. Variable cadence: 5 s in far mode, 1 s in close mode\n', 'DELTA_PLUS_VAR': 'DELTA_EPOCH', 'CADENCE': '1 second, 5 seconds', 'FIELDNAM': 'Epoch', 'CATDESC': 'Nanoseconds since J2000'}

Variable: DELTA_EPOCH
Attributes (attrs): {'FILLVAL': np.uint32(4294967295), 'LABLAXIS': 'Delta epoch', 'UNITS': 'seconds', 'VALIDMIN': np.uint32(0), 'VALIDMAX': np.uint32(3600), 'SCALEMIN': np.uint32(0), 'SCALEMAX': np.uint32(3600), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': '', 'FIELDNAM': 'Delta epoch', 'CATDESC': 'Accumulation period in seconds', 'DEPEND_0': 'EPOCH', 'FORMAT': 'I'}

Variable: Ion_Flux
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'LABLAXIS': 'Ion Flux', 'UNITS': 'particles / (s cm^2 sr MeV)', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(1e+20), 'SCALEMIN': np.float32(3.213347), 'SCALEMAX': np.float32(12070.784), 'VAR_TYPE': 'data', 'SCALETYP': 'log', 'VAR_NOTES': 'Detector response can be different for different ion species. The calibrated intensity has been obtained using the response for Hydrogen.\n', 'DELTA_PLUS_VAR': 'Ion_Uncertainty', 'FIELDNAM': 'Ion Flux', 'CATDESC': 'Particle flux in magnet channel', 'DEPEND_0': 'EPOCH', 'FORMAT': 'G10.4', 'DEPEND_1': 'Ion_Bins_Low_Energy', 'DISPLAY_TYPE': 'spectrogram', 'SI_CONVERSION': '6.2415e16>particles/(s m^2 sr J)', 'DELTA_MINUS_VAR': 'Ion_Uncertainty', 'LABL_PTR_1': 'Ion_Bins_Text'}

Variable: Ion_Uncertainty
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'LABLAXIS': 'Ion Flux Sigma', 'UNITS': 'particles / (s cm^2 sr MeV)', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(1e+20), 'SCALEMIN': np.float32(3.213347), 'SCALEMAX': np.float32(6969.071), 'VAR_TYPE': 'support_data', 'SCALETYP': 'log', 'VAR_NOTES': '', 'FIELDNAM': 'Ion Flux Uncertainty', 'CATDESC': 'Uncertainty in ion flux', 'DEPEND_0': 'EPOCH', 'FORMAT': 'G10.4', 'DEPEND_1': 'Ion_Bins_Low_Energy', 'DISPLAY_TYPE': 'spectrogram', 'SI_CONVERSION': '6.2415e16>particles/(s m^2 sr J)'}

Variable: Ion_Rate
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'LABLAXIS': 'Ion Count Rate', 'UNITS': 'counts / s', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(1e+20), 'SCALEMIN': np.float32(0.1), 'SCALEMAX': np.float32(3.0), 'VAR_TYPE': 'support_data', 'SCALETYP': 'log', 'VAR_NOTES': '', 'FIELDNAM': 'Ion Count Rate', 'CATDESC': 'Particle count rate in magnet channel', 'DEPEND_0': 'EPOCH', 'FORMAT': 'G10.4', 'DEPEND_1': 'Ion_Bins_Low_Energy', 'DISPLAY_TYPE': 'spectrogram', 'LABL_PTR_1': 'Ion_Bins_Text'}

Variable: Ion_Bins_Low_Energy
Attributes (attrs): {'LABLAXIS': 'Energy', 'UNITS': 'MeV', 'VALIDMIN': np.float32(1e-06), 'VALIDMAX': np.float32(1000000000.0), 'SCALEMIN': np.float32(0.05175), 'SCALEMAX': np.float32(4.099), 'VAR_TYPE': 'support_data', 'SCALETYP': 'log', 'VAR_NOTES': '', 'DELTA_PLUS_VAR': 'Ion_Bins_Width', 'FIELDNAM': 'Ion Bins Low Energy', 'CATDESC': 'Lowest energy of each bin for ion flux', 'FORMAT': 'G10.4', 'SI_CONVERSION': '1.602e-13>J', 'LABL_PTR_1': 'Ion_Bins_Text'}

Variable: Ion_Bins_Width
Attributes (attrs): {'LABLAXIS': 'Bin width', 'UNITS': 'MeV', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(1000000000.0), 'VAR_TYPE': 'support_data', 'VAR_NOTES': '', 'FIELDNAM': 'Ion Bins Width', 'CATDESC': 'Width of each energy bin for ion flux', 'FORMAT': 'G10.4', 'SI_CONVERSION': '1.602e-13>J'}

Variable: Ion_Bins_Text
Attributes (attrs): {'VAR_TYPE': 'metadata', 'VAR_NOTES': '', 'FIELDNAM': 'Ion Bins Text', 'CATDESC': 'Text description of each energy bin for ion flux', 'FORMAT': 'A25'}

Variable: Electron_Flux
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'LABLAXIS': 'Electron Flux', 'UNITS': 'particles / (s cm^2 sr MeV)', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(1e+20), 'SCALEMIN': np.float32(112.044815), 'SCALEMAX': np.float32(45687.137), 'VAR_TYPE': 'data', 'SCALETYP': 'log', 'VAR_NOTES': 'Ions with energies around 500 keV can penetrate the foil and give a signal in the electron channel. In the presence of high ion flux, there could be a considerable contamination from ions in this product.\n', 'DELTA_PLUS_VAR': 'Electron_Uncertainty', 'FIELDNAM': 'Electron Flux', 'CATDESC': 'Particle flux in foil channel', 'DEPEND_0': 'EPOCH', 'FORMAT': 'G10.4', 'DEPEND_1': 'Electron_Bins_Low_Energy', 'DISPLAY_TYPE': 'spectrogram', 'SI_CONVERSION': '6.2415e16>particles/(s m^2 sr J)', 'DELTA_MINUS_VAR': 'Electron_Uncertainty', 'LABL_PTR_1': 'Electron_Bins_Text'}

Variable: Electron_Uncertainty
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'LABLAXIS': 'Electron Flux Sigma', 'UNITS': 'particles / (s cm^2 sr MeV)', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(1e+20), 'SCALEMIN': np.float32(112.044815), 'SCALEMAX': np.float32(32305.684), 'VAR_TYPE': 'support_data', 'SCALETYP': 'log', 'VAR_NOTES': '', 'FIELDNAM': 'Electron Flux Uncertainty', 'CATDESC': 'Uncertainty in electron flux', 'DEPEND_0': 'EPOCH', 'FORMAT': 'G10.4', 'DEPEND_1': 'Electron_Bins_Low_Energy', 'DISPLAY_TYPE': 'spectrogram', 'SI_CONVERSION': '6.2415e16>particles/(s m^2 sr J)'}

Variable: Electron_Rate
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'LABLAXIS': 'Electron Count Rate', 'UNITS': 'counts / s', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(1e+20), 'SCALEMIN': np.float32(0.1), 'SCALEMAX': np.float32(2.0), 'VAR_TYPE': 'support_data', 'SCALETYP': 'log', 'VAR_NOTES': 'Affected by contamination from ions with energies ~ 500 keV', 'FIELDNAM': 'Electron Count Rate', 'CATDESC': 'Particle count rate in foil channel', 'DEPEND_0': 'EPOCH', 'FORMAT': 'G10.4', 'DEPEND_1': 'Electron_Bins_Low_Energy', 'DISPLAY_TYPE': 'spectrogram', 'LABL_PTR_1': 'Electron_Bins_Text'}

Variable: Electron_Bins_Low_Energy
Attributes (attrs): {'LABLAXIS': 'Energy', 'UNITS': 'MeV', 'VALIDMIN': np.float32(1e-06), 'VALIDMAX': np.float32(1000000000.0), 'SCALEMIN': np.float32(0.03186), 'SCALEMAX': np.float32(0.3994), 'VAR_TYPE': 'support_data', 'SCALETYP': 'log', 'VAR_NOTES': '', 'DELTA_PLUS_VAR': 'Electron_Bins_Width', 'FIELDNAM': 'Electron Bins Low Energy', 'CATDESC': 'Lowest energy of each bin for electron flux', 'FORMAT': 'G10.4', 'SI_CONVERSION': '1.602e-13>J', 'LABL_PTR_1': 'Electron_Bins_Text'}

Variable: Electron_Bins_Width
Attributes (attrs): {'LABLAXIS': 'Bin width', 'UNITS': 'MeV', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(1000000000.0), 'VAR_TYPE': 'support_data', 'VAR_NOTES': '', 'FIELDNAM': 'Electron Bins Width', 'CATDESC': 'Width of each energy bin for electron flux', 'FORMAT': 'G10.4', 'SI_CONVERSION': '1.602e-13>J'}

Variable: Electron_Bins_Text
Attributes (attrs): {'VAR_TYPE': 'metadata', 'VAR_NOTES': '', 'FIELDNAM': 'Electron Bins Text', 'CATDESC': 'Text description of each energy bin for electron flux', 'FORMAT': 'A25'}

Variable: EPOCH_1
Attributes (attrs): {'FILLVAL': np.int64(-9223372036854775808), 'LABLAXIS': 'Time', 'UNITS': 'nanoseconds', 'VALIDMIN': np.int64(-43135816000000), 'VALIDMAX': np.int64(1577880069184000000), 'SCALEMIN': np.int64(647870469184000000), 'SCALEMAX': np.int64(647956869184000000), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': '', 'CADENCE': '60 seconds', 'FIELDNAM': 'Epoch 1', 'CATDESC': 'Nanoseconds since J2000'}

Variable: EPOCH_2
Attributes (attrs): {'FILLVAL': np.int64(-9223372036854775808), 'LABLAXIS': 'Time', 'UNITS': 'nanoseconds', 'VALIDMIN': np.int64(-43135816000000), 'VALIDMAX': np.int64(1577880069184000000), 'SCALEMIN': np.int64(647870469184000000), 'SCALEMAX': np.int64(647956869184000000), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': '', 'CADENCE': '1 hour', 'FIELDNAM': 'Epoch 2', 'CATDESC': 'Nanoseconds since J2000'}

Variable: RTN
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'UNITS': ' ', 'VALIDMIN': np.float32(-1.0), 'VALIDMAX': np.float32(1.0), 'SCALEMIN': np.float32(-1.0), 'SCALEMAX': np.float32(1.0), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': '', 'FIELDNAM': 'Flow direction RTN', 'CATDESC': 'Particle flow direction (unit vector) in RTN coordinates', 'DEPEND_0': 'EPOCH_1', 'FORMAT': 'F6.3', 'DISPLAY_TYPE': 'time_series', 'LABL_PTR_1': 'RTN_Labels'}

Variable: XYZ
Attributes (attrs): {'UNITS': ' ', 'VALIDMIN': np.float32(-1.0), 'VALIDMAX': np.float32(1.0), 'VAR_TYPE': 'support_data', 'VAR_NOTES': '', 'FIELDNAM': 'Flow direction XYZ', 'CATDESC': 'Particle flow direction (unit vector) in spacecraft XYZ coordinates', 'FORMAT': 'F6.3', 'LABL_PTR_1': 'XYZ_Labels'}

Variable: HCI_R
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'LABLAXIS': 'r', 'UNITS': 'au', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(1.2), 'SCALEMIN': np.float32(0.0), 'SCALEMAX': np.float32(1.2), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': '', 'FIELDNAM': 'S/C Radial Distance', 'CATDESC': 'Spacecraft radial distance from the Sun', 'DEPEND_0': 'EPOCH_2', 'FORMAT': 'F6.4', 'SI_CONVERSION': '1.495978E11>m'}

Variable: HCI_Lat
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'LABLAXIS': 'Latitude', 'UNITS': 'degrees', 'VALIDMIN': np.float32(-90.0), 'VALIDMAX': np.float32(90.0), 'SCALEMIN': np.float32(-90.0), 'SCALEMAX': np.float32(90.0), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': '', 'FIELDNAM': 'S/C Latitude', 'CATDESC': 'Spacecraft heliocentric latitude (HCI)', 'DEPEND_0': 'EPOCH_2', 'FORMAT': 'F6.2'}

Variable: HCI_Lon
Attributes (attrs): {'FILLVAL': np.float32(-1e+31), 'LABLAXIS': 'Longitude', 'UNITS': 'degrees', 'VALIDMIN': np.float32(0.0), 'VALIDMAX': np.float32(360.0), 'SCALEMIN': np.float32(0.0), 'SCALEMAX': np.float32(360.0), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': '', 'FIELDNAM': 'S/C Longitude', 'CATDESC': 'Spacecraft heliocentric longitude (HCI)', 'DEPEND_0': 'EPOCH_2', 'FORMAT': 'F6.2'}

Variable: QUALITY_FLAG
Attributes (attrs): {'FILLVAL': np.uint8(255), 'LABLAXIS': 'Quality flag', 'UNITS': ' ', 'VALIDMIN': np.uint8(0), 'VALIDMAX': np.uint8(4), 'SCALEMIN': np.uint8(0), 'SCALEMAX': np.uint8(4), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': '0:Bad, 1:Known problems, 2:Survey data, 3:Good, 4:Excellent', 'FIELDNAM': 'Quality flag', 'CATDESC': 'Data quality flag', 'DEPEND_0': 'EPOCH', 'FORMAT': 'I'}

Variable: QUALITY_BITMASK
Attributes (attrs): {'FILLVAL': np.uint16(65535), 'LABLAXIS': 'Quality bitmask', 'UNITS': ' ', 'VALIDMIN': np.uint16(0), 'VALIDMAX': np.uint16(32767), 'SCALEMIN': np.uint16(0), 'SCALEMAX': np.uint16(32767), 'VAR_TYPE': 'support_data', 'SCALETYP': 'linear', 'VAR_NOTES': '0x1: Bad value, 0x2: Close mode, 0x4: High flux mode\n', 'FIELDNAM': 'Quality bitmask', 'CATDESC': 'Computer readable quality parameter', 'DEPEND_0': 'EPOCH', 'FORMAT': 'Z4'}

Variable: RTN_Labels
Attributes (attrs): {'VAR_TYPE': 'metadata', 'FIELDNAM': 'RTN Labels', 'CATDESC': 'Labels for vector components in RTN coordinates', 'FORMAT': 'A1'}

Variable: XYZ_Labels
Attributes (attrs): {'VAR_TYPE': 'metadata', 'FIELDNAM': 'Spacecraft XYZ Labels', 'CATDESC': 'Labels for vector components in spacecraft XYZ coordinates', 'FORMAT': 'A1'}

this is the attrs dictionary i got from the test cdf file so 

these variables contains FILLVAL key those who contains FILLVAL i have converted them to NaN while none FILLVAL keys are as it is.

sh-abinash-sh replied:
```
def test_generic_timeseries_columns():
    # Read the CDF file using cdflib to get variable attributes
    cdf = cdflib.CDF(filepath)
    var_keys = cdf.cdf_info().rVariables + cdf.cdf_info().zVariables
    var_attrs = {key: cdf.varattsget(key) for key in var_keys}
    
    # Identify variables with 'FILLVAL'
    fillval_vars = {key: attrs['FILLVAL'] for key, attrs in var_attrs.items() if 'FILLVAL' in attrs}
    
    # Read the data using read_cdf
    all_ts = read_cdf(filepath)
    assert isinstance(all_ts, list)
    assert len(all_ts) >= 1  # Ensure at least one time series is returned
    
    # Process each GenericTimeSeries
    for ts in all_ts:
        assert isinstance(ts, GenericTimeSeries)
        
        # Map CDF variable keys to column names in GenericTimeSeries
        # Assume that column names are either the variable key or var_key_i for multi-column variables
        for col_name in ts.columns:
            # Determine the original variable key
            if '_' in col_name:
                var_key = '_'.join(col_name.split('_')[:-1])
            else:
                var_key = col_name
            
            # Skip if the variable key is not in var_attrs (可能被跳过)
            if var_key not in var_attrs:
                continue
            
            # Get the data and its attributes
            col_data = ts.quantity(col_name)
            attrs = var_attrs[var_key]
            
            # Check based on 'FILLVAL' presence
            if var_key in fillval_vars:
                # Original fill value
                fillval = fillval_vars[var_key]
                
                # Get original data from CDF
                original_data = cdf.varget(var_key)
                
                # For multi-column variables, extract the specific column
                if original_data.ndim == 2:
                    col_index = int(col_name.split('_')[-1])
                    original_col_data = original_data[:, col_index]
                else:
                    original_col_data = original_data
                
                # Positions where original data had fillval
                fill_positions = original_col_data == fillval
                
                # Ensure these positions are NaN in the column data
                assert np.all(np.isnan(col_data[fill_positions]))
                
                # Optionally, ensure no additional NaNs are introduced
                # This depends on the data; might not always be true
                # non_fill_positions = ~fill_positions
                # assert np.sum(np.isnan(col_data[non_fill_positions])) == np.sum(np.isnan(original_col_data[non_fill_positions]))
            else:
                # Ensure no NaNs are introduced unless already present
                # Check if original data had NaNs
                original_data = cdf.varget(var_key)
                
                if original_data.ndim == 2:
                    col_index = int(col_name.split('_')[-1])
                    original_col_data = original_data[:, col_index]
                else:
                    original_col_data = original_data
                
                # If original data did not have NaNs, the column data should not have NaNs
                if not np.any(np.isnan(original_col_data)):
                    assert not np.any(np.isnan(col_data))
                else:
                    # If original data had NaNs, ensure they are preserved
                    nan_positions = np.isnan(original_col_data)
                    assert np.all(np.isnan(col_data[nan_positions]))
```

This is the code to do that.

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -23,3 +25,83 @@ def test_read_cdf():
     assert np.sum(np.isnan(col)) == 189
+
+
+def test_generic_timeseries_columns():
```
All we need to check is that the column of the timeseries which should have nans in, has nans in it.

nabobalis replied:
We should also check that the original code example passes now as well.

sh-abinash-sh commented:
![Screenshot 2024-12-12 203913](https://github.com/user-attachments/assets/1affbe94-b2e9-43e2-b0ed-c336926e10c5)

checked the column of the timeseries which should have nans in, has nans in it.

this also passes the original code.
![Screenshot 2024-12-12 204207](https://github.com/user-attachments/assets/6dfdb7a0-35e9-4a14-af9a-504917c899e0)

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -23,3 +23,23 @@ def test_read_cdf():
     assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
     # Check that fillvals are replaced by NaN
     assert np.sum(np.isnan(col)) == 189
+
+def test_check_nan_values():
+    all_ts = read_cdf(filepath)
+    assert isinstance(all_ts, list)
+
+    assert len(all_ts) > 0
+
+    for i, ts in enumerate(all_ts):
+        print(f"Checking Timeseries {i + 1} for NaN values...")
+        has_nan = False
+
+        data = ts.to_dataframe()
+        for column in data.columns:
+            if data[column].isna().any():
+                has_nan = True
+                print(f"  Column '{column}' contains NaN values.")
+
+        if not has_nan:
+            print(f"  No NaN values found in Timeseries {i + 1}.")
+        print()
```
This isn't how a test should be written. Please have a read over https://realpython.com/pytest-python-testing/

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -23,3 +23,11 @@ def test_read_cdf():
     assert np.sum(np.isnan(col)) == 189
+
+
+def test_check_nan_values():
```
Did this test fail before the code change?

sh-abinash-sh replied:
No before it was working fine , now after the latest commit it is failing

sh-abinash-sh replied:
![Screenshot 2024-12-13 000016](https://github.com/user-attachments/assets/eeefb742-7829-42a5-8f48-a351cc379a53)

nabobalis replied:
That needs to be fixed but that was not what I was referring to. 

The goal behind the unit test is to make sure that we test the original bug is fixed.

If this test passes before and after the bugfix, this test is useless. 

We need a unit test that fails before the bugfix change and then will pass after said change.

Can you write one?

sh-abinash-sh commented:
Now it should pass , for 1st time time series columns it should be ```first_ts = read_cdf(filepath)[0]``` not first_ts = read_cdf(filepath)[1] if we change this than we have to check about column name 'RTN_0', 'RTN_1', 'RTN_2'.

nabobalis commented:
> ![Screenshot 2024-12-13 000511](https://private-user-images.githubusercontent.com/191173990/395286846-e03f4977-48fc-43f8-b41b-2a58a56230c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwMjk5MDcsIm5iZiI6MTczNDAyOTYwNywicGF0aCI6Ii8xOTExNzM5OTAvMzk1Mjg2ODQ2LWUwM2Y0OTc3LTQ4ZmMtNDNmOC1iNDFiLTJhNThhNTYyMzBjMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxMlQxODUzMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNmQ0N2M5MzJhOTQ5NzI2NGEwYzQ0MDMyYzRhN2IxYmE0OTBjMzkzYmQ1YTBiMzc2YTg4YjRlNDkwNmEzYWJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.33USsDUH35V8faCPGyKxLZ-wzaj0zVTVPpeAOqM1NL8) Now it should pass , for 1st time time series columns it should be `first_ts = read_cdf(filepath)[0]` not `first_ts = read_cdf(filepath)[1]` if we change this than we have to check about column name 'RTN_0', 'RTN_1', 'RTN_2'.

The goal behind the unit test is to make sure that we test the original bug is fixed.

If this test passes before and after the bugfix, this test is useless.

We need a unit test that fails before the bugfix change and then will pass after said change.

Can you write one?

sh-abinash-sh commented:
> > ![Screenshot 2024-12-13 000511](https://private-user-images.githubusercontent.com/191173990/395286846-e03f4977-48fc-43f8-b41b-2a58a56230c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwMjk5MDcsIm5iZiI6MTczNDAyOTYwNywicGF0aCI6Ii8xOTExNzM5OTAvMzk1Mjg2ODQ2LWUwM2Y0OTc3LTQ4ZmMtNDNmOC1iNDFiLTJhNThhNTYyMzBjMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxMlQxODUzMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNmQ0N2M5MzJhOTQ5NzI2NGEwYzQ0MDMyYzRhN2IxYmE0OTBjMzkzYmQ1YTBiMzc2YTg4YjRlNDkwNmEzYWJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.33USsDUH35V8faCPGyKxLZ-wzaj0zVTVPpeAOqM1NL8) Now it should pass , for 1st time time series columns it should be `first_ts = read_cdf(filepath)[0]` not `first_ts = read_cdf(filepath)[1]` if we change this than we have to check about column name 'RTN_0', 'RTN_1', 'RTN_2'.
> 
> The goal behind the unit test is to make sure that we test the original bug is fixed.
> 
> If this test passes before and after the bugfix, this test is useless.
> 
> We need a unit test that fails before the bugfix change and then will pass after said change.
> 
> Can you write one?

Please suggest me any guidance.

I got it like before and after it is same the tests are passing before and as well as now .
What should i do now.

nabobalis commented:
> Please suggest me any guidance.
> 
> I got it like before and after it is same the tests are passing before and as well as now . What should i do now.

Simplest answer is to use the code example in the original issue.

sh-abinash-sh commented:
> > Please suggest me any guidance.
> > I got it like before and after it is same the tests are passing before and as well as now . What should i do now.
> 
> Simplest answer is to use the code example in the original issue.

```
def test_read_psp_data():
    # Define the dataset and time range
    dataset = 'PSP_SWP_SPI_SF00_L3_MOM'
    trange = a.Time(dt.date(2023, 3, 14), dt.date(2023, 3, 15))

    # Search and fetch the data
    result = Fido.search(trange, a.cdaweb.Dataset(dataset))
    downloaded_files = Fido.fetch(result)

    ts = TimeSeries(downloaded_files, concatenate=True)

    # Read the first downloaded file
    # data = read_cdf(downloaded_files.data[0])
    print(ts.columns)
    assert isinstance(ts, TimeSeries), "The data should be a list of TimeSeries objects."

    col = ts.quantity('EFLUX_VS_ENERGY_0')
    assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
    assert np.sum(np.isnan(col)) == 189
```

i have written this new test function based upon the original issue code:

But while running the tests:

i got an unexpected connection error

![Screenshot 2024-12-13 094137](https://github.com/user-attachments/assets/b356221f-3f37-4b02-a68b-b5b7c9445d9f)

nabobalis commented:
> > > Please suggest me any guidance.
> > > I got it like before and after it is same the tests are passing before and as well as now . What should i do now.
> > 
> > 
> > Simplest answer is to use the code example in the original issue.
> 
> ```
> def test_read_psp_data():
>     # Define the dataset and time range
>     dataset = 'PSP_SWP_SPI_SF00_L3_MOM'
>     trange = a.Time(dt.date(2023, 3, 14), dt.date(2023, 3, 15))
> 
>     # Search and fetch the data
>     result = Fido.search(trange, a.cdaweb.Dataset(dataset))
>     downloaded_files = Fido.fetch(result)
> 
>     ts = TimeSeries(downloaded_files, concatenate=True)
> 
>     # Read the first downloaded file
>     # data = read_cdf(downloaded_files.data[0])
>     print(ts.columns)
>     assert isinstance(ts, TimeSeries), "The data should be a list of TimeSeries objects."
> 
>     col = ts.quantity('EFLUX_VS_ENERGY_0')
>     assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
>     assert np.sum(np.isnan(col)) == 189
> ```
> 
> i have written this new test function based upon the original issue code:
> 
> But while running the tests:
> 
> i got an unexpected connection error
> 
> ![Screenshot 2024-12-13 094137](https://private-user-images.githubusercontent.com/191173990/395423945-b356221f-3f37-4b02-a68b-b5b7c9445d9f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwNjM5NzcsIm5iZiI6MTczNDA2MzY3NywicGF0aCI6Ii8xOTExNzM5OTAvMzk1NDIzOTQ1LWIzNTYyMjFmLTNmMzctNGIwMi1hNjhiLWI1YjdjOTQ0NWQ5Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxM1QwNDIxMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZjExMGRjOTM4ODZmY2FmNTNlYTBlNjNkMGY0MzMyZjllMjdkZWU1YmZlZjViY2E2NjNlYTE0MGJmNDc0OTA3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jPGSCjCqNIrWrs_FLzp5FZOvcZj6KvQcbB1Wop86a6A)

See https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html#online-tests

sh-abinash-sh commented:
> > > > Please suggest me any guidance.
> > > > I got it like before and after it is same the tests are passing before and as well as now . What should i do now.
> > > 
> > > 
> > > Simplest answer is to use the code example in the original issue.
> > 
> > 
> > ```
> > def test_read_psp_data():
> >     # Define the dataset and time range
> >     dataset = 'PSP_SWP_SPI_SF00_L3_MOM'
> >     trange = a.Time(dt.date(2023, 3, 14), dt.date(2023, 3, 15))
> > 
> >     # Search and fetch the data
> >     result = Fido.search(trange, a.cdaweb.Dataset(dataset))
> >     downloaded_files = Fido.fetch(result)
> > 
> >     ts = TimeSeries(downloaded_files, concatenate=True)
> > 
> >     # Read the first downloaded file
> >     # data = read_cdf(downloaded_files.data[0])
> >     print(ts.columns)
> >     assert isinstance(ts, TimeSeries), "The data should be a list of TimeSeries objects."
> > 
> >     col = ts.quantity('EFLUX_VS_ENERGY_0')
> >     assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
> >     assert np.sum(np.isnan(col)) == 189
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > i have written this new test function based upon the original issue code:
> > But while running the tests:
> > i got an unexpected connection error
> > ![Screenshot 2024-12-13 094137](https://private-user-images.githubusercontent.com/191173990/395423945-b356221f-3f37-4b02-a68b-b5b7c9445d9f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwNjM5NzcsIm5iZiI6MTczNDA2MzY3NywicGF0aCI6Ii8xOTExNzM5OTAvMzk1NDIzOTQ1LWIzNTYyMjFmLTNmMzctNGIwMi1hNjhiLWI1YjdjOTQ0NWQ5Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxM1QwNDIxMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZjExMGRjOTM4ODZmY2FmNTNlYTBlNjNkMGY0MzMyZjllMjdkZWU1YmZlZjViY2E2NjNlYTE0MGJmNDc0OTA3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jPGSCjCqNIrWrs_FLzp5FZOvcZj6KvQcbB1Wop86a6A)
> 
> See https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html#online-tests

the test is getting skipped while i have an established internet connection even after using the decorator.

nabobalis commented:
> > > > > Please suggest me any guidance.
> > > > > I got it like before and after it is same the tests are passing before and as well as now . What should i do now.
> > > > 
> > > > 
> > > > Simplest answer is to use the code example in the original issue.
> > > 
> > > 
> > > ```
> > > def test_read_psp_data():
> > >     # Define the dataset and time range
> > >     dataset = 'PSP_SWP_SPI_SF00_L3_MOM'
> > >     trange = a.Time(dt.date(2023, 3, 14), dt.date(2023, 3, 15))
> > > 
> > >     # Search and fetch the data
> > >     result = Fido.search(trange, a.cdaweb.Dataset(dataset))
> > >     downloaded_files = Fido.fetch(result)
> > > 
> > >     ts = TimeSeries(downloaded_files, concatenate=True)
> > > 
> > >     # Read the first downloaded file
> > >     # data = read_cdf(downloaded_files.data[0])
> > >     print(ts.columns)
> > >     assert isinstance(ts, TimeSeries), "The data should be a list of TimeSeries objects."
> > > 
> > >     col = ts.quantity('EFLUX_VS_ENERGY_0')
> > >     assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
> > >     assert np.sum(np.isnan(col)) == 189
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > i have written this new test function based upon the original issue code:
> > > But while running the tests:
> > > i got an unexpected connection error
> > > ![Screenshot 2024-12-13 094137](https://private-user-images.githubusercontent.com/191173990/395423945-b356221f-3f37-4b02-a68b-b5b7c9445d9f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwNjM5NzcsIm5iZiI6MTczNDA2MzY3NywicGF0aCI6Ii8xOTExNzM5OTAvMzk1NDIzOTQ1LWIzNTYyMjFmLTNmMzctNGIwMi1hNjhiLWI1YjdjOTQ0NWQ5Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxM1QwNDIxMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZjExMGRjOTM4ODZmY2FmNTNlYTBlNjNkMGY0MzMyZjllMjdkZWU1YmZlZjViY2E2NjNlYTE0MGJmNDc0OTA3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jPGSCjCqNIrWrs_FLzp5FZOvcZj6KvQcbB1Wop86a6A)
> > 
> > 
> > See [docs.sunpy.org/en/latest/dev_guide/contents/tests.html#online-tests](https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html#online-tests)
> 
> the test is getting skipped while i have an established internet connection even after using the decorator.

Then you need to try: https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html#remote-data

sh-abinash-sh commented:
![Screenshot 2024-12-13 101903](https://github.com/user-attachments/assets/5ac2c39c-1b63-41ef-a283-2a28ee0cb754)
the test fails before bug fix

sh-abinash-sh commented:
![Screenshot 2024-12-13 102048](https://github.com/user-attachments/assets/3f71bf76-9edd-4ab4-a142-068115b0e642)
While after bug fix i got this warning converting to an error

nabobalis commented:
> ![Screenshot 2024-12-13 102048](https://private-user-images.githubusercontent.com/191173990/395432837-3f71bf76-9edd-4ab4-a142-068115b0e642.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQwNjU5MTIsIm5iZiI6MTczNDA2NTYxMiwicGF0aCI6Ii8xOTExNzM5OTAvMzk1NDMyODM3LTNmNzFiZjc2LTllZGQtNGFiNC1hMTQyLTA2ODExNWIwZTY0Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxM1QwNDUzMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Zjk4ZDJjYmVkMzkzYzI2NTIzM2Y1NGQxODBjMDc5MjFiNmFjYzlmOWM4ZGY1MjViMTdlNDJlNzBhYjAwOTc2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.V3OzbL3EEEt1t5bUC8PpQ1bZGAzbhD1LwjDxpOX5uA8) While after bug fix i got this warning converting to an error

You will need to suppress or catch the warning via pytest.

sh-abinash-sh commented:
Warning for units in cdf
    ignore: Something went wrong: astropy did not recognize units of "eV/cm2-s-ster-eV". Assigning dimensionless units. If you think this unit should not be dimensionless, please raise an issue at https://github.com/sunpy/sunpy/issues :sunpy.util.exceptions.SunpyUserWarning
    
I wrote in pytest.ini this but i got this issue:

![Screenshot 2024-12-13 104103](https://github.com/user-attachments/assets/48627f66-308d-450f-8e52-38c0d5cb51c1)

sh-abinash-sh commented:
i tried to supress the warning in pytest.ini by writting this but i still the error remains same.

 Warning for units in cdf
    ignore:Something went wrong astropy did not recognize units of.*:sunpy.util.exceptions.SunpyUserWarning

nabobalis commented:
> i tried to supress the warning in pytest.ini by writting this but i still the error remains same.
> 
> Warning for units in cdf ignore:Something went wrong astropy did not recognize units of.*:sunpy.util.exceptions.SunpyUserWarning

pytest.ini should only be used if the problem is widespread. This should be handled inside the test function.

sh-abinash-sh commented:
```
@pytest.mark.filterwarnings("ignore:SunpyUserWarning:")
def test_read_psp_data():

    # Define the dataset and time range
    dataset = 'PSP_SWP_SPI_SF00_L3_MOM'
    trange = a.Time(dt.date(2023, 3, 14), dt.date(2023, 3, 15))

    # Search and fetch the data
    result = Fido.search(trange, a.cdaweb.Dataset(dataset))
    downloaded_files = Fido.fetch(result)

    ts = TimeSeries(downloaded_files, concatenate=True)

    # Read the first downloaded file
    # data = read_cdf(downloaded_files.data[0])
    print(ts.columns)
    assert isinstance(ts, TimeSeries), "The data should be a list of TimeSeries objects."

    col = ts.quantity('EFLUX_VS_ENERGY_0')
    assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
    assert np.sum(np.isnan(col)) == 189
```

I used the decorators to supress the warning but it still remians same.

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -23,3 +27,26 @@ def test_read_cdf():
     assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
     # Check that fillvals are replaced by NaN
     assert np.sum(np.isnan(col)) == 189
+
+@pytest.mark.remote_data
+def test_read_psp_data():
+
+    # Define the dataset and time range
+    dataset = 'PSP_SWP_SPI_SF00_L3_MOM'
+    trange = a.Time(dt.date(2023, 3, 14), dt.date(2023, 3, 15))
+
+    # Search and fetch the data
+    result = Fido.search(trange, a.cdaweb.Dataset(dataset))
+    downloaded_files = Fido.fetch(result)
+
+    ts = TimeSeries(downloaded_files, concatenate=True)
+
+    # Read the first downloaded file
+    # data = read_cdf(downloaded_files.data[0])
+    print(ts.columns)
+    assert isinstance(ts, GenericTimeSeries), "The data should be a list of TimeSeries objects."
+
+    col = ts.quantity('EFLUX_VS_ENERGY_0')
+    # assert col.unit == u.Unit("1 / (cm2 MeV s sr)")
+    print(col)
+    assert np.sum(np.isnan(col)) >= 0
```
```suggestion


@pytest.mark.remote_data
def test_read_psp_data():
    # This was a failing example provided by
    # https://github.com/sunpy/sunpy/issues/7565
    dataset = 'PSP_SWP_SPI_SF00_L3_MOM'
    trange = a.Time(dt.date(2023, 3, 14), dt.date(2023, 3, 15))
    result = Fido.search(trange, a.cdaweb.Dataset(dataset))
    downloaded_files = Fido.fetch(result)
    ts = TimeSeries(downloaded_files, concatenate=True)
    assert isinstance(ts, GenericTimeSeries)
    col = ts.quantity('EFLUX_VS_ENERGY_0')
    assert np.sum(np.isnan(col)) >= 0
```

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -1,16 +1,20 @@
 
 filepath = get_test_filepath('solo_L2_epd-ept-north-hcad_20200713_V02.cdf')
 
-
```
This line needs to be added back

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -1,16 +1,20 @@
 
-
 def test_read_cdf():
+
```
This line needs to be removed.

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -1,16 +1,20 @@
+import datetime as dt
```
Why do we need this?

sh-abinash-sh replied:
this is needed to extract trange in 2nd test function.

nabobalis replied:
Why not use strings?

sh-abinash-sh replied:
Done

sh-abinash-sh commented:
CI / docs / build_docs (ubuntu-latest) (pull_request) failing!

nabobalis commented on changelog/7917.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Updated the internal CDF reader to handle `FILLVAL`` only for floating point numbers.
```
```suggestion
Updated the internal CDF reader to handle ``FILLVAL`` only for floating point numbers.
```

nabobalis commented:
> CI / docs / build_docs (ubuntu-latest) (pull_request) failing!

Yes, well spotted.

Abinash-bit commented:
All checks passes now.

sh-abinash-sh commented:
All check passes!

nabobalis commented:
```
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/io/tests/test_cdf.py::test_read_psp_data - sunpy.util.exceptions.SunpyUserWarning: Something went wrong: astropy did not recognize units of "eV/cm2-s-ster-eV". Assigning dimensionless units. If you think this unit should not be dimensionless, please raise an issue at [https://github.com/sunpy/sunpy/issues](https://github.com/sunpy/sunpy/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen)
```
Looks like the error was not caught.

sh-abinash-sh commented:
It will pass now.

sh-abinash-sh commented:
All checks finally!!

nabobalis commented on sunpy/io/tests/test_cdf.py:
```diff
@@ -23,3 +28,20 @@ def test_read_cdf():
+    result = Fido.search(trange, a.cdaweb.Dataset(dataset))
+    downloaded_files = Fido.fetch(result)
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", category=SunpyUserWarning)
```
You can not catch with with pytest.warns?

sh-abinash-sh replied:
can not understand

nabobalis replied:
Maybe this will help https://docs.pytest.org/en/stable/how-to/capture-warnings.html ?

sh-abinash-sh replied:
done it through by pytest.warns.

nabobalis replied:
Can you match part of the message?

sh-abinash-sh replied:
Done i guess so.!

nabobalis replied:
I dont see the message match?

sh-abinash-sh commented:
Now it should be ok.!

nabobalis commented:
Thanks @sh-abinash-sh
